syncthing/syncthing
# Add WebAuthn support to GUI (fixes #8409)

Fixes #8409.

### Remaining things to do

- [x] Rewrite data type changes in protobuf
- [x] Store and validate signature counter
- [ ] ~~Store transports and return in `allowCredentials`~~ Not currently supported by [duo-labs/webauthn](https://github.com/duo-labs/webauthn)
- [x] Remove backwards-incompatible template string syntax in JS
- [x] Remove basic auth prompt
- [x] Proper password auth API endpoint
- [x] WebAuthn failure messages on login page
- [x] Warning on login page when on raw IP address
- [ ] Migrate from deprecated duo-labs/webauthn to successor fork
- [ ] Open documentation PR

### Purpose

This adds support for [WebAuthn](https://webauthn.guide/) authentication in the GUI. WebAuthn is a W3C standard that provides an API for strong, privacy-conscious public-key authentication, which can enable a both more pleasant and more secure user experience than traditional password authentication. WebAuthn both works with external security keys (typically via USB or NFC) and is built into many modern browsers and operating systems.

For users who already use WebAuthn, this both cuts out a detour through a password manager and improves security by making it possible to not need a password at all.


### Testing

Manual tests performed:

- With no WebAuthn credentials enrolled:
  - If no password set, no authentication is required
  - ~~Basic authorization prompt appears like before~~ Removed!
- With WebAuthn credentials enrolled:
  - With password set, Basic authorization headers still work as before
  - An HTML login page with both WebAuthn and password options is presented
  - Password authentication in HTML form works (no WebAuthn required)
  - WebAuthn authentication in HTML form works (no password required)
  - WebAuthn authentication is available immediately without server restart
- WebAuthn settings/credential management:
  - Warning appears when HTTPS is disabled
  - Warning appears when RP ID does not match page domain
  - Warning appears when WebAuthn origin does not match RP ID
  - Warning appears when changing RP ID
  - New credentials appear immediately in the table in GUI settings
  - Renaming credentials takes effect on saving the GUI settings
  - Credential names fall back to credential ID when not set
  - Deleting credentials takes effect on saving the GUI settings
  - `PUT /rest/config` ignores WebAuthn credentials not already registered (identified by ID)
  - [`excludeCredentials`][exclude] is used to prevent registering the same authenticator twice

I intend to add automated tests as well, but would first like a review of the overall design and how it fits into the existing architecture.


### Screenshots

![HTML login form](https://user-images.githubusercontent.com/1367758/177319721-a6903abc-598f-48f6-a85c-1e7981676735.png)
HTML login form

![New GUI settings](https://user-images.githubusercontent.com/1367758/177319989-85b9ffc8-f308-414b-8676-08ab9f2ee840.png)
New GUI settings

![Registering a new credential](https://user-images.githubusercontent.com/1367758/177320105-8a1a211e-2be9-4b73-9351-a6c663a0f41f.png)
Registering a new credential

![Renaming a credential](https://user-images.githubusercontent.com/1367758/177320169-fb4c0336-b06b-4348-90a4-4645f805cd1a.png)
Renaming a credential

![Interactive warnings](https://user-images.githubusercontent.com/1367758/177320206-8c9e1abb-6fc1-40e0-8d31-7a2b00231ad6.png)
Interactive warnings


### Documentation

Probably a new section should be added alongside "LDAP Authentication". WebAuthn normally doesn't require much technical knowledge from the user since you only interact with the end-user side, but in the Syncthing case you are both end-user and server administrator. WebAuthn requires some domain settings to line up, so the "RP ID" and "WebAuthn Origin" settings and their impact need to be explained here. I volunteer to write this documentation once I have the green light to finish the feature.


[exclude]: https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-excludecredentials
[transports]: https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-transports
[signCount]: https://w3c.github.io/webauthn/#signcount
